### ISSUE TITLE
com.amazonaws.services.s3.model.AmazonS3Exception: The specified key does not exist.

### ISSUE THREAD

**Issue by atarsdnh** [awslabs/aws-sdk-android-samples#127]
I am trying to use TransferUtility to upload and download files.

I replace the following code:
```
GetObjectRequest gor = 
                new GetObjectRequest(Constants.BUCKET_NAME, key);
        s3Client.getObject(gor, new File(destinationFilePath));
```
to:
```
sTransferUtility.download(
                Constants.BUCKET_NAME,
                key,
                new File(destinationFilePath),
                listener);
```

And then I meet the Exception. I guess the reason is my key contains "@", but I don't know how to handle it to match the bucket file, how do I download and upload file with key name contains "@" by TransferUtility?

**Comment by rohandubal** [comment]
Can you please give some more details to help us better?

-What is the version of SDK you are using?
-Does the request succeed for getObject and fail for transfer utility?
-Are you uploading the object using putObject / transfer utility?

Thanks,
Rohan

**Comment by atarsdnh** [comment]
My SDK version:
```
    compile 'com.amazonaws:aws-android-sdk-core:2.2.+'
    compile 'com.amazonaws:aws-android-sdk-s3:2.2.+'
```

Yes, the request succeed for getObject and fail for transfer utility with same key.

If upload the object using putObject with key "test@gmail.com/image.jpg" works fine, but using transfer utility with the same key, the file in the bucket will transfer to test%40gmail.com/image.jpg.
I think download using transfer utility fail because I use "test@gmail.com/image.jpg" as key but there is no "test%40gmail.com/image.jpg" file.

**Comment by rohandubal** [comment]
2.2 is a very old SDK version. Could you update to latest 2.4.2 and verify if the issue is still existing? The encoding issue would likely be fixed in the newer SDK.

Thanks,
Rohan

**Comment by atarsdnh** [comment]
I tested 2.2.+ and 2.4.+ SDK version with same code using transfer utility:

If I use 2.2.+ to upload with key name "test@gmail.com/image.jpg", the file path will be "test@gmail.com/image.jpg", but 2.4.+ will encoding to "test%40gmail.com/image.jpg".

How do I access the "test@gmail.com/image.jpg" file if I update to the latest SDK version?

**Comment by rohandubal** [comment]
The file should be `test@gmail.com/image.jpg` after being uploaded to S3. So with 2.2.+ it works correctly but 2.4.+ it doesn't? I will try reproducing this on my side.

**Comment by maxbax** [comment]
I think have the same error. I uploaded and downloaded asset correctly until version 2.3.9. After this version (I tried with the latest 2.4.4 too) seems that the key is encoded before upload. 
If I upload with a key without special char (a space only for example) it works, but if I write a space in the key, for example, the library upload the file into bucket, but I can't download it with the same key. 

I've tried to rename the file uploaded into bucket replacing the "%20" with spaces and I downloaded it again with the url I used (https://.....s3.amazonaws.com/u/0/126/Mm....4/48016/Filename%200.....). So I think it could be a problem with a probably encode of key during upload
